### DISCUSSION TITLE
OpenCV 介紹分享文

### DISCUSSION THREAD

**Discussion by DuxtinTai** [1101-GS4719-C/Python-Community-GS4719-C#1266] [優秀Python套件]
首先，OpenCV是甚麼? Open Source Computer Vision Library!
OpenCV是一個強大的計算機視覺庫，可應用於多個領域。它具有圖像處理、影像處理、物件識別、視頻處理、機器學習等多種功能，能夠實現圖像或視頻的處理、分析和重建。另外，OpenCV支持多種編程語言，如C++、Python等，並提供了豐富的函數庫和算法，方便開發者使用。其開源的特性也吸引了眾多開發者的參與，使其得到了不斷的發展和完善。

詳述功能如下:
1. 圖像處理：OpenCV提供了多種圖像處理函數，如圖像濾波、二值化、邊緣檢測、形態學運算等，可以實現圖像的增強、修復、分割等操作。OpenCV還提供了多種圖像轉換函數，如幾何變換、色彩空間轉換等。
2. 特徵檢測：OpenCV提供了多種特徵檢測算法，如Harris角點檢測、SIFT和SURF等，可以用於圖像的配準、物體檢測和跟蹤等。OpenCV還提供了多種描述符算法，如BRIEF和ORB等，可以將圖像特徵描述為向量，便於匹配和識別。
3. 目標檢測和識別：OpenCV提供了多種目標檢測和識別算法，如Haar特徵分類器、深度學習算法等，可以實現人臉識別、車輛識別等應用。OpenCV還提供了多種機器學習算法，如支持向量機、決策樹和隨機森林等，可用於圖像分類和回歸等任務。
4. 三維重建：OpenCV提供了多種三維重建算法，如立體視覺、結構光等，可以實現物體三維形狀的恢復和測量等應用。OpenCV還提供了多種三維可視化函數，可以將三維重建結果以圖像或視頻的形式呈現。
5. 視頻處理：OpenCV提供了多種視頻處理函數，如幀差分、光流估計等，可以實現視頻的穩定、去噪、分割等操作。OpenCV還提供了多種視頻編解碼器和格式轉換函數，可用於視頻的壓縮和轉換。
6. 機器學習：OpenCV提供了多種機器學習函數和算法，如支持向量機、決策樹和隨機森林等，可用於圖像分類和回歸等任務。OpenCV還支持深度學習框架，如TensorFlow和Caffe等，可以實現更複雜的圖像識別和分類任務。

接下來，我想分享我之前的經驗。在我的專題中，我使用了OpenCV來進行電路板標示絲印模擬的視覺化，以應用於標明電路板上各元件的位置。透過OpenCV的圖像製作功能，我們能夠輕鬆辨識出電路板上繪製的絲印和各元件的位置，進而驗證絲印標示的位置是否正確。從實際應用的角度來看，OpenCV的語法直觀易懂，自由度高，擁有豐富的功能，是繪圖方面的好幫手。

**Comment by chia-kai-chang** [comment]
> 從實際應用的角度來看，OpenCV的語法直觀易懂，自由度高，擁有豐富的功能，是繪圖方面的好幫手。

是繪圖的好幫手? 他好像是做影像辨識，不是繪圖ㄋㄝ~

**Reply by DuxtinTai** [reply]
OpenCV主要的應用的確是在圖像辨識，這也是他最POWERFUL的地方，但是他也有繪圖功能喔~
<img width="220" alt="image" src="https://user-images.githubusercontent.com/124947223/233651939-847c1bac-bfe6-460c-8091-8d9cbf17fb7e.png">
這張圖就是用opencv製作出來的，透過給定圖紙大小，像素，再加上線段或弧線位置、顏色等就可以製作出圖片喔
像是這張圖黑線部分是電路板的主體部分，綠色線就是銅箔的位置，橘色線則是我們設計的絲印位置~
當然，opencv除了可以用在繪圖方面外，也可以用實體電路板的位置偵測喔，只是這塊我們沒有用到

**Reply by HsuSungTing** [reply]
太電了吧 Daxting
直接拿特優的作品來Comment
現在的Comment品質都這麼高嗎XD

**Comment by DuxtinTai** [comment]
這邊再來分享一下用於繪圖的幾個tips，繪圖最需要注意的是在opencv中最左上角才是原點位置(0,0)，向下是y軸的正向，向右是x軸的正向，這邊稍微比較反直覺。
簡易繪圖的教學連結:
https://opencv-python-tutorials.readthedocs.io/zh/latest/2.%20OpenCV%E4%B8%AD%E7%9A%84%20Gui%E7%89%B9%E6%80%A7/2.3.%20OpenCV%E4%B8%AD%E7%9A%84%E7%BB%98%E5%9B%BE%E5%8A%9F%E8%83%BD/
另外，前面有提到opencv支持多種語言，如果想使用c++在vs上使用opencv函式庫，推薦下方連結的教學，很清楚!
https://kaibaoom.tw/2019/07/18/opencv-vs2019/
